### PR TITLE
Inital LND rpc with some tests

### PR DESCRIPTION
**Pull request by benthecarman**
Still have lots of functions to give official methods as well as needing to add more extensive tests. However, would like some intial review

### CODE REVIEW

File: .github/workflows/Linux_2.12_RPC_Tests.yml [hunk 1]
```diff
@@ -27,4 +27,4 @@ jobs:
             ~/.bitcoin-s/binaries
           key: ${{ runner.os }}-cache
       - name: run tests
-        run: sbt ++2.12.12 downloadBitcoind downloadEclair coverage bitcoindRpcTest/test bitcoindRpc/coverageReport bitcoindRpc/coverageAggregate bitcoindRpc/coveralls eclairRpcTest/test eclairRpc/coverageReport eclairRpc/coverageAggregate eclairRpc/coveralls
\ No newline at end of file
+        run: sbt ++2.12.12 downloadBitcoind downloadEclair downloadLnd coverage bitcoindRpcTest/test bitcoindRpc/coverageReport bitcoindRpc/coverageAggregate bitcoindRpc/coveralls eclairRpcTest/test eclairRpc/coverageReport eclairRpc/coverageAggregate eclairRpc/coveralls lndRpcTest/test
```
Comment (Christewart) [inline .github/workflows/Linux_2.12_RPC_Tests.yml:30]: Can we change the name on this CI matrix above to add lnd

File: lnd-rpc/src/main/scala/org/bitcoins/lnd/rpc/config/LndConfig.scala [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+    }.headOption
+  }
+
+  /** Converts a string to an InetSocketAddress */
+  private def toInetSocketAddress(string: String): InetSocketAddress = {
+    val uri = new URI(string)
+    new InetSocketAddress(uri.getHost, uri.getPort)
+  }
+
+  lazy val bitcoindUser: String = getValue("bitcoind.rpcuser").getOrElse("foo")
```
Comment (Christewart) [inline lnd-rpc/src/main/scala/org/bitcoins/lnd/rpc/config/LndConfig.scala]: Probably should just use `.get` so that we throw in the case it isn't defined.

Let's fail fast rather than use a bad config that probably will fail else where.

File: lnd-rpc/src/main/scala/org/bitcoins/lnd/rpc/config/LndConfig.scala [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+
+object LndConfig extends Logging {
+
+  /** The empty `lnd` config */
+  lazy val empty: LndConfig = LndConfig("", DEFAULT_DATADIR)
+
+  /** Constructs a `lnd` config from the given string,
+    * by splitting it on newlines
+    */
+  def apply(config: String, datadir: File): LndConfig =
```
Comment (Christewart) [inline lnd-rpc/src/main/scala/org/bitcoins/lnd/rpc/config/LndConfig.scala]: Can we get some sort of common interface across all this? I believe this is the same with bitcoind/eclair too 

https://github.com/bitcoin-s/bitcoin-s/blob/f593cf063e53b2c8e6b867d5c975d824694b31cd/bitcoind-rpc/src/main/scala/org/bitcoins/rpc/config/BitcoindInstance.scala#L123

Comment (Christewart) [reply lnd-rpc/src/main/scala/org/bitcoins/lnd/rpc/config/LndConfig.scala]: This is sort of similar, but is meant for `AppConfig` https://github.com/bitcoin-s/bitcoin-s/blob/ee7c96245e99325d1386464c4462d36dce7bcafb/db-commons/src/main/scala/org/bitcoins/db/AppConfigFactory.scala#L9

File: lnd-rpc/src/main/scala/org/bitcoins/lnd/rpc/config/LndConfig.scala [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+        s"We will not overwrite the existing lnd.conf in default datadir")
+    } else {
+      Files.write(confFile, confStr.getBytes)
+    }
+
+    confFile
+  }
+}
+
+sealed abstract class LogLevel
```
Comment (Christewart) [inline lnd-rpc/src/main/scala/org/bitcoins/lnd/rpc/config/LndConfig.scala]: Maybe this can be moved somewhere else?

File: lnd-rpc/src/main/scala/org/bitcoins/lnd/rpc/config/LndInstance.scala [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+        val bytes = Files.readAllBytes(path)
+        val hex = ByteVector(bytes).toHex
+
+        macaroonOpt = Some(hex)
+        hex
+    }
+  }
+}
+
+object LndInstance {
```
Comment (Christewart) [inline lnd-rpc/src/main/scala/org/bitcoins/lnd/rpc/config/LndInstance.scala]: Same thing here, it would be nice to have a common interface. 

I guess we don't have to figure it all out on this PR, at minimum though we should file an issue

File: lnd-rpc/src/main/scala/org/bitcoins/lnd/rpc/LndRpcClient.scala [hunk 6]
```diff
@@ -0,0 +1,517 @@
...
+      s"$file --lnddir=${instance.datadir.toAbsolutePath}"
+    case None => ""
+  }
+
+  implicit val executionContext: ExecutionContext = system.dispatcher
+
+  // These need to be lazy so we don't try and fetch
+  // the tls certificate before it is generated
+
+  lazy val certStream = new FileInputStream(instance.certFile)
```
Comment (Christewart) [inline lnd-rpc/src/main/scala/org/bitcoins/lnd/rpc/LndRpcClient.scala]: Can some of these things be `private`? 

We don't want to expose some of these to users of `LndRpcClient` IMO

File: lnd-rpc/src/main/scala/org/bitcoins/lnd/rpc/LndRpcClient.scala [hunk 7]
```diff
@@ -0,0 +1,517 @@
...
+import java.io.{File, FileInputStream}
+import java.net.InetSocketAddress
+import java.util.concurrent.atomic.AtomicInteger
+import scala.concurrent.duration.{DurationInt, FiniteDuration}
+import scala.concurrent.{ExecutionContext, Future, Promise}
+import scala.util.{Failure, Success, Try}
+
+/** @param binary Path to lnd executable
+  */
+class LndRpcClient(val instance: LndInstance, binary: Option[File] = None)(
```
Comment (Christewart) [inline lnd-rpc/src/main/scala/org/bitcoins/lnd/rpc/LndRpcClient.scala:35]: Back to API discussion, is there some subset of APIs that are common across Eclair and Lnd? 

We could create some sort of `LightningRpcApi` that is extended by both this and `EclairRpcClient`. 

That way, when building lightning applications we can be more abstract rather than coding to a specific implementation. 

Here is `EclairApi` 

https://github.com/bitcoin-s/bitcoin-s/blob/5b762fb2e13742f057162e14a035b67730053ca0/eclair-rpc/src/main/scala/org/bitcoins/eclair/rpc/api/EclairApi.scala#L29

File: lnd-rpc/lnd-rpc.sbt [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+name := "bitcoin-s-lnd-rpc"
+
+libraryDependencies ++= Deps.lndRpc
+
+CommonSettings.prodSettings
+
+enablePlugins(AkkaGrpcPlugin)
+
+// Disable fatal warning otherwise protobuf deprecation warnings will cause errors
+scalacOptions in Compile ~= (_ filterNot (s => s == "-Xfatal-warnings"))
```
Comment (Christewart) [inline lnd-rpc/lnd-rpc.sbt]: You should be able to filter this by using the same tools I used in this PR (specifically ignoring errors in the `protobuf` package)

https://github.com/bitcoin-s/bitcoin-s/pull/2822

File: testkit/src/main/scala/org/bitcoins/testkit/fixtures/LndFixture.scala [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+import org.bitcoins.lnd.rpc.LndRpcClient
+import org.bitcoins.rpc.client.common.BitcoindRpcClient
+import org.bitcoins.rpc.client.v21.BitcoindV21RpcClient
+import org.bitcoins.testkit.lnd.{LndRpcTestClient, LndRpcTestUtil}
+import org.bitcoins.testkit.rpc.BitcoindRpcTestUtil
+import org.bitcoins.testkit.util.BitcoinSAsyncFixtureTest
+import org.scalatest.FutureOutcome
+
+/** A trait that is useful if you need Lnd fixtures for your test suite */
+trait LndFixture extends BitcoinSFixture {
```
Comment (Christewart) [inline testkit/src/main/scala/org/bitcoins/testkit/fixtures/LndFixture.scala]: We should definitely try and incorporate `CachedBitcoind` in here otherwise `withLnd()`/`withDualLnd` will spin up fresh bitcoinds/lnds for every test case. 

see https://github.com/bitcoin-s/bitcoin-s/pull/2792

Happy to help as this is newer code in our testkit project, and it is using some more advanced scala features

File: bitcoind-rpc/src/main/scala/org/bitcoins/rpc/util/NativeProcessFactory.scala [hunk 10]
```diff
@@ -33,8 +33,10 @@ trait NativeProcessFactory extends Logging {
         logger.debug(s"Binary was already started!")
         ()
       case None =>
-        val started = process.run()
-        processOpt = Some(started)
+        if (cmd.nonEmpty) {
```
Comment (Christewart) [inline bitcoind-rpc/src/main/scala/org/bitcoins/rpc/util/NativeProcessFactory.scala:36]: maybe put a warning log in the `else` clause?

**Review by Christewart — CHANGES_REQUESTED** [review]

File: lnd-rpc/lnd-rpc.sbt (hunk 8 above)
Comment (benthecarman) [reply lnd-rpc/lnd-rpc.sbt]: Doesn't seem you can disable deprecation warnings for a single package

File: lnd-rpc/src/main/scala/org/bitcoins/lnd/rpc/LndRpcClient.scala (hunk 7 above)
Comment (benthecarman) [reply lnd-rpc/src/main/scala/org/bitcoins/lnd/rpc/LndRpcClient.scala:35]: This sounds good in theory but don't think will work out to well. These 2 return and take in a lot of different types for matching functions (ie eclair returns a `FundedChannelId` for openChannel and lnd returns a `Option[TransactionOutpoint]`) so I don't know how useful we could actually make it

**Comment by benthecarman** [comment]
Note: fix one of `App, Chain, Node, and Core Tests` running forever

File: lnd-rpc/src/main/scala/org/bitcoins/lnd/rpc/LndRpcClient.scala (hunk 7 above)
Comment (Christewart) [reply lnd-rpc/src/main/scala/org/bitcoins/lnd/rpc/LndRpcClient.scala:35]: I'll file an issue to think about this further. We can make their API conform to our API in some cases, but the example you cite is a good one, if `None` is given to us we can't translate that into a `FundedChannelId` :/

File: lnd-rpc/lnd-rpc.sbt (hunk 8 above)
Comment (Christewart) [reply lnd-rpc/lnd-rpc.sbt]: Here is the incantation you need

```scala
scalacOptions in Compile += {
  //Added to supress all of the deprecation warnings inside of the
  //generated scala code by the grpc plugin. Two packages have deprecations
  //inside of them, 'signrpc' and 'lnrpc'
  //see: https://www.scala-lang.org/2021/01/12/configuring-and-suppressing-warnings.html
  if (scalaVersion.value.startsWith("2.13")) {
    "-Wconf:cat=deprecation:site=lnrpc\\..*:silent,cat=deprecation:site=signrpc\\..*:silent"
  } else {
    //-Wconf doesn't work on Scala 2.12.x until we get 2.12.13
    //which is currently blocked on a scoverage bug
    ""
  }
}
```

File: lnd-rpc/src/main/scala/org/bitcoins/lnd/rpc/LndRpcClient.scala (hunk 7 above)
Comment (Christewart) [reply lnd-rpc/src/main/scala/org/bitcoins/lnd/rpc/LndRpcClient.scala:35]: https://github.com/bitcoin-s/bitcoin-s/issues/2844

File: testkit/src/main/scala/org/bitcoins/testkit/fixtures/LndFixture.scala (hunk 9 above)
Comment (benthecarman) [reply testkit/src/main/scala/org/bitcoins/testkit/fixtures/LndFixture.scala]: I think I did this correctly

File: testkit/src/main/scala/org/bitcoins/testkit/fixtures/LndFixture.scala [hunk 11]
```diff
@@ -19,30 +18,28 @@ trait LndFixture extends BitcoinSFixture {
...
       }
     )(test)
   }
 }
 
 /** A trait that is useful if you need Lnd fixtures for your test suite */
-trait DualLndFixture extends BitcoinSFixture {
-  _: BitcoinSAsyncFixtureTest =>
+trait DualLndFixture extends BitcoinSFixture with CachedBitcoindV21 {
+  _: BitcoinSAsyncFixtureTest with CachedBitcoindFunded[_] =>
```
Comment (Christewart) [inline testkit/src/main/scala/org/bitcoins/testkit/fixtures/LndFixture.scala]: So in plain english this is what this means
```scala
trait DualLndFixture extends BitcoinSFixture with CachedBitcoindV21 {
  _: BitcoinSAsyncFixtureTest with CachedBitcoindFunded[_] =>
```

>I've created a trait called `DualLndFixture` that extends `BitcoinSFixture with CachedBitcoindV21`. It _requires_ a `BitcoinSAsyncFixtureTest` AND `CachedBitcoindFunded[_]` to be injected into this trait before it can be used.

This doesn't make much sense, because you _already_ have a `CachedBitcoindV21` extended, so why do we need to inject _another_ `CachedBitcoindFunded[_]`. Its a bit redundant. 

you probably want to do something like this

```scala
trait DualLndFixture extends BitcoinSFixture with BitcoindFixturesFundedCachedV21
``` 

and then i don't think you need any `{_: BitcoinSAsyncFixtureTest with CachedBitcoindFunded[_] =>` 

Although I didn't run this locally to test it myself

File: testkit/src/main/scala/org/bitcoins/testkit/fixtures/LndFixture.scala [hunk 12]
```diff
@@ -54,19 +51,18 @@ trait DualLndFixture extends BitcoinSFixture {
     makeDependentFixture[FixtureParam](
       () => {
         for {
-          bitcoind <- LndRpcTestUtil.startedBitcoindRpcClient()
+          bitcoind <- cachedBitcoindWithFundsF
           _ = logger.debug("starting bitcoind")
           _ <- bitcoind.start()
```
Comment (Christewart) [inline testkit/src/main/scala/org/bitcoins/testkit/fixtures/LndFixture.scala]: Shouldn't need to be started if you are using a cached bitcoind

**Comment by benthecarman** [comment]
Re-ran lnd rpc ci a bunch and haven't gotten any failures

File: app-commons/src/main/scala/org/bitcoins/commons/ConfigFactory.scala [hunk 13]
```diff
@@ -0,0 +1,28 @@
+package org.bitcoins.commons
+
+import java.io.File
+import java.nio.file.Path
+
+trait ConfigFactory[T] {
```
Comment (Christewart) [inline app-commons/src/main/scala/org/bitcoins/commons/ConfigFactory.scala]: nit: I don't think there is any reason this can't be moved into `org.bitcoins.core.api.commons`

File: app-commons/src/main/scala/org/bitcoins/commons/InstanceFactory.scala [hunk 14]
```diff
@@ -0,0 +1,13 @@
+package org.bitcoins.commons
+
+import java.io.File
+import java.nio.file.Path
+
+trait InstanceFactory[+T] {
```
Comment (Christewart) [inline app-commons/src/main/scala/org/bitcoins/commons/InstanceFactory.scala]: same comment wrt to moving to `core`

File: testkit/src/main/scala/org/bitcoins/testkit/lnd/LndRpcTestUtil.scala [hunk 15]
```diff
@@ -0,0 +1,344 @@
...
+  }
+
+  def lndInstance(datadir: File): LndInstance = {
+    LndInstance.fromDataDir(datadir)
+  }
+
+  /** Returns a `Future` that is completed when both lnd and bitcoind have the same block height
+    * Fails the future if they are not synchronized within the given timeout.
+    */
+  def awaitLndInSync(lnd: LndRpcClient, bitcoind: BitcoindRpcClient)(implicit
```
Comment (Christewart) [inline testkit/src/main/scala/org/bitcoins/testkit/lnd/LndRpcTestUtil.scala:146]: Not related to this PR, but I wanted to leave a note to re-visit this when we address #2844 

It would be very nice to have abstractions for this so we don't have to write new helper methods to do it

Comment (Christewart) [reply testkit/src/main/scala/org/bitcoins/testkit/lnd/LndRpcTestUtil.scala:146]: Really the rest of the file

**Review by Christewart — APPROVED** [review]
